### PR TITLE
Add a script to count the lines that are pending resubmission

### PR DESCRIPTION
**Pull request by marcusmueller**
as of now:

```
missing: "andy@silverblocksystems.net" (1 LOC)
missing: "douggeiger@users.noreply.github.com" (1 LOC)
missing: "gnieboer@gcndevelopment.com" (1 LOC)
missing: "jamarck96@gmail.com" (1 LOC)
missing: "julien.olivain@lsv.ens-cachan.fr" (1 LOC)
missing: "steven_behnke@me.com" (1 LOC)
missing: "jam_quiceno@partech.com" (3 LOC)
missing: "stefan-oltmanns@gmx.net" (3 LOC)
missing: "doron.behar@gmail.com" (4 LOC)
missing: "n.corgan@gmail.com" (5 LOC)
missing: "gwenhael.goavec-merou@trabucayre.com" (6 LOC)
missing: "adamt@nvidia.com" (7 LOC)
missing: "alexandre.rouma@gmail.com" (9 LOC)
missing: "ben.hilburn@ettus.com" (9 LOC)
missing: "brandon.enochs@nrl.navy.mil" (9 LOC)
missing: "pascal.giard@lacime.etsmtl.ca" (9 LOC)
missing: "qwerty15@gmx.fr" (10 LOC)
missing: "luzpaz@users.noreply.github.com" (11 LOC)
missing: "mrene@avgs.ca" (11 LOC)
missing: "nate.ewest@gmail.com" (11 LOC)
missing: "val.zapod.vz@gmail.com" (11 LOC)
missing: "nick.corgan@ettus.com" (16 LOC)
missing: "dan@dlrobertson.com" (22 LOC)
missing: "rthompso@gmail.com" (26 LOC)
missing: "tnt@246tNt.com" (26 LOC)
missing: "carlesfernandez@gmail.com" (30 LOC)
missing: "damian.miralles@colorado.edu" (30 LOC)
missing: "alesha72003@ya.ru" (35 LOC)
missing: "gnieboer@corpcomm.net" (41 LOC)
missing: "tim.oshea753@gmail.com" (46 LOC)
missing: "nathan.west@nrl.navy.mil" (60 LOC)
missing: "aholguin_77@yahoo.com" (62 LOC)
missing: "Christoph.Mayer@cern.ch" (81 LOC)
missing: "johnathan@corganlabs.com" (94 LOC)
missing: "dmiralles2009@gmail.com" (99 LOC)
missing: "elliot.briggs@gmail.com" (180 LOC)
missing: "nick@ettus.com" (182 LOC)
missing: "hcab14@gmail.com" (342 LOC)
missing: "nick@nerdnetworks.org" (358 LOC)
missing: "ritterho@hm.edu" (366 LOC)
missing: "nwest@deepsig.io" (472 LOC)
missing: "damian.miralles@aero.org" (527 LOC)
missing: "nathan.west@gnuradio.org" (538 LOC)
missing: "abhowmick22@gmail.com" (553 LOC)
missing: "trondeau@vt.edu" (713 LOC)
missing: "carles.fernandez@gmail.com" (1015 LOC)
missing: "jessica.iwamoto@aero.org" (2653 LOC)
missing: "josh@joshknows.com" (5306 LOC)
missing: "nathan.west@okstate.edu" (7414 LOC)
missing: "tom@trondeau.com" (7452 LOC)
Missing 28863 of 70404 LOC in total (40.99%)
```

### CODE REVIEW

**Comment by marcusmueller** [comment]
missing: "andy@silverblocksystems.net" (1 LOC)
missing: "douggeiger@users.noreply.github.com" (1 LOC)
missing: "gnieboer@gcndevelopment.com" (1 LOC)
missing: "jamarck96@gmail.com" (1 LOC)
missing: "julien.olivain@lsv.ens-cachan.fr" (1 LOC)
missing: "steven_behnke@me.com" (1 LOC)
missing: "jam_quiceno@partech.com" (3 LOC)
missing: "stefan-oltmanns@gmx.net" (3 LOC)
missing: "doron.behar@gmail.com" (4 LOC)
missing: "n.corgan@gmail.com" (5 LOC)
missing: "gwenhael.goavec-merou@trabucayre.com" (6 LOC)
missing: "adamt@nvidia.com" (7 LOC)
missing: "alexandre.rouma@gmail.com" (9 LOC)
missing: "ben.hilburn@ettus.com" (9 LOC)
missing: "brandon.enochs@nrl.navy.mil" (9 LOC)
missing: "pascal.giard@lacime.etsmtl.ca" (9 LOC)
missing: "qwerty15@gmx.fr" (10 LOC)
missing: "luzpaz@users.noreply.github.com" (11 LOC)
missing: "mrene@avgs.ca" (11 LOC)
missing: "nate.ewest@gmail.com" (11 LOC)
missing: "val.zapod.vz@gmail.com" (11 LOC)
missing: "nick.corgan@ettus.com" (16 LOC)
missing: "dan@dlrobertson.com" (22 LOC)
missing: "rthompso@gmail.com" (26 LOC)
missing: "tnt@246tNt.com" (26 LOC)
missing: "carlesfernandez@gmail.com" (30 LOC)
missing: "damian.miralles@colorado.edu" (30 LOC)
missing: "alesha72003@ya.ru" (35 LOC)
missing: "gnieboer@corpcomm.net" (41 LOC)
missing: "tim.oshea753@gmail.com" (46 LOC)
missing: "nathan.west@nrl.navy.mil" (60 LOC)
missing: "aholguin_77@yahoo.com" (62 LOC)
missing: "Christoph.Mayer@cern.ch" (81 LOC)
missing: "johnathan@corganlabs.com" (94 LOC)
missing: "dmiralles2009@gmail.com" (99 LOC)
missing: "elliot.briggs@gmail.com" (180 LOC)
missing: "nick@ettus.com" (182 LOC)
missing: "hcab14@gmail.com" (342 LOC)
missing: "nick@nerdnetworks.org" (358 LOC)
missing: "ritterho@hm.edu" (366 LOC)
missing: "nwest@deepsig.io" (472 LOC)
missing: "damian.miralles@aero.org" (527 LOC)
missing: "nathan.west@gnuradio.org" (538 LOC)
missing: "abhowmick22@gmail.com" (553 LOC)
missing: "trondeau@vt.edu" (713 LOC)
missing: "carles.fernandez@gmail.com" (1015 LOC)
missing: "jessica.iwamoto@aero.org" (2653 LOC)
missing: "josh@joshknows.com" (5306 LOC)
missing: "nathan.west@okstate.edu" (7414 LOC)
missing: "tom@trondeau.com" (7452 LOC)
Missing 28863 of 70446 LOC in total (40.97%)

File: scripts/licensing/count_contrib.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#/bin/bash
```
Comment (jdemel) [inline scripts/licensing/count_contrib.sh]: At the moment, this is a hard requirement. The script runs in `bash` but not `sh` (I assume `dash` on Ubuntu 20.04) or `zsh`.

File: scripts/licensing/count_contrib.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# relicense due to their contributions being under an acceptable license
+# already" if we need; no changes to this script would be necessary.
+#
+# This script is part of VOLK.
+# 
+# Copyright 2021 Marcus Müller
+# SPDX-License-Identifier: MPL-2.0
+
+rootdir=`git rev-parse --show-toplevel`
+if [[ "$#" -lt 1 ]]
```
Comment (jdemel) [inline scripts/licensing/count_contrib.sh:17]: I replaced the `[[` with `[` to allow `zsh` and `sh` to run the script.

File: scripts/licensing/count_contrib.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+then
+    authorfile=$rootdir/AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md
+else
+    authorfile=$1
+fi
+if [[ ! -r $authorfile ]]
```
Comment (jdemel) [inline scripts/licensing/count_contrib.sh:23]: Same `[[` -> `[`.
Otherwise: `[[: not found`

File: scripts/licensing/count_contrib.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    then
+        echo "missing: \"$authoremail\" (${authorlines} LOC)"
+        missingloc=$(($missingloc + $authorlines))
+    fi
+done < <(echo "$authorcounts")
+
+percentage=$(echo "scale=2; 100.0 * $missingloc/$total_loc" | bc)
+echo "Missing $missingloc of $total_loc LOC in total ($percentage%)"
+
+if [[  "$missingloc" -gt 0 ]]
```
Comment (jdemel) [inline scripts/licensing/count_contrib.sh:51]: Same `[[` -> `[`.

File: scripts/licensing/count_contrib.sh [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+do
+    authoremail=$(echo "$line" | sed 's/^ *\([[:digit:]]*\) *\([^, ]*\)$/\2/g')
+    authorlines=$(echo "$line" | sed 's/^ *\([[:digit:]]*\) *\([^, ]*\)$/\1/g')
+    total_loc=$(( $authorlines + $total_loc ))
+    if ! ( echo "$lgplers" | grep -i "$authoremail" ) > /dev/null
+    then
+        echo "missing: \"$authoremail\" (${authorlines} LOC)"
+        missingloc=$(($missingloc + $authorlines))
+    fi
+done < <(echo "$authorcounts")
```
Comment (jdemel) [inline scripts/licensing/count_contrib.sh:46]: That line breaks without `bash`. Do you have some reference what is done here and how to use this?

**Review by jdemel — COMMENTED** [review]
Thanks for that script! I really like it!
Still, I have a few comments. I hope you can help me out here.

File: scripts/licensing/count_contrib.sh (hunk 1 above)
Comment (marcusmueller) [reply scripts/licensing/count_contrib.sh]: yes. No intent to change that; this is nothing that needs to be portable to systems without bash. And I'm using enough command substitutions and control structures to not want to go for full sh compat anyway.

File: scripts/licensing/count_contrib.sh (hunk 2 above)
Comment (marcusmueller) [reply scripts/licensing/count_contrib.sh:17]: I'd respectfully point out that this isn't something to be run on systems without bash.

File: scripts/licensing/count_contrib.sh (hunk 3 above)
Comment (marcusmueller) [reply scripts/licensing/count_contrib.sh:23]: yes, run the script with the program specified in the shebang line: You'd not expect me to write Py2k-compatible Python, either!

File: scripts/licensing/count_contrib.sh (hunk 5 above)
Comment (marcusmueller) [reply scripts/licensing/count_contrib.sh:46]: I'm using command substitution to feed the output of echo into the `read`. Again, this script relies on bash and porting it to sh is more work than it's worth.

Comment (jdemel) [reply scripts/licensing/count_contrib.sh:46]: True. I'll just invoke it with `bash ./....`. I marked all the "breaks outside bash" comments as resolved.

Still, could you point me to a tutorial that explains this exact line? I'd like to learn how to use it.

**Comment by marcusmueller** [comment]
Your feedback for converting this to `sh` is very appreciated, but I simply don't have the time and drive to do that! It's interesting some bashes fail to understand the `<()`, whereas mine does just fine; hm, that's a bug. Does it run cleanly for you locally, @jdemel ?

File: CMakeLists.txt [hunk 6]
```diff
@@ -423,6 +423,13 @@ if(ENABLE_MODTOOL)
   add_subdirectory(python/volk_modtool)
 endif()
 
+########################################################################
+# And the LGPL license check
+########################################################################
+if(ENABLE_TESTING)
+  add_test("check_lgpl" ${CMAKE_SOURCE_DIR}/scripts/licensing/count_contrib.sh ${CMAKE_SOURCE_DIR}/AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md)
```
Comment (jdemel) [inline CMakeLists.txt:438]: I really like the fact that we check for LGPL in a test. I suggest to add another test though. 
This solution breaks all tests at the moment.

**Comment by marcusmueller** [comment]
D'oh. Forgot the bang!

**Comment by jdemel** [comment]
> Your feedback for converting this to `sh` is very appreciated, but I simply don't have the time and drive to do that! It's interesting some bashes fail to understand the `<()`, whereas mine does just fine; hm, that's a bug. Does it run cleanly for you locally, @jdemel ?

Now it runs without hiccups. I can just do `./scripts/licensing/count_contrib.sh` in any shell too. The shebang fix helped a lot.

**Comment by marcusmueller** [comment]
made the `check_lgpl` sensitive to `cmake -DENABLE_LGPL=On/Off`, with the default being Off, as long as the major version number not >= 3.

File: scripts/licensing/count_contrib.sh (hunk 5 above)
Comment (marcusmueller) [reply scripts/licensing/count_contrib.sh:46]: https://www.gnu.org/software/bash/manual/html_node/Process-Substitution.html

Comment (jdemel) [reply scripts/licensing/count_contrib.sh:46]: Thanks!
Just for the sake of completeness:
https://www.gnu.org/software/bash/manual/html_node/Process-Substitution.html
explains how to use `<(...)`.
And
https://linuxize.com/post/bash-while-loop/#read-a-file-line-by-line
Explains how to use the while loop.

**Review by jdemel — COMMENTED** [review]
Thanks! LGTM. One last thing. Some commits miss the `Signed-off-by` line.

**Review by jdemel — APPROVED** [review]
LGTM. Thanks!

**Review by michaelld — APPROVED** [review]
LGTM ; thx!

**Comment by michaelld** [comment]
I'm fairly confident that we'll need to tweak this script & maybe the CMake code once we start using it more; we just can't up-front know all of the idiosyncratic use cases that we'll encounter. Maybe I'm wrong; will always hope this is good as is!

**Comment by michaelld** [comment]
@marcusmueller would you please squash & DCO & force push?

**Comment by marcusmueller** [comment]
I'll try my best, but I'm currently moving; squashing should make the dco error go away, so if anyone feels like squashing/rebasing and force-pushing, I won't mind... 

On October 8, 2021 4:12:14 PM GMT+02:00, Michael Dickens ***@***.***> wrote:
***@***.*** would you please squash & DCO & force push? 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/gnuradio/volk/pull/526#issuecomment-938676082
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by jdemel** [comment]
@marcusmueller I would've squashed these commits for you but that'd take away the purpose why we use DCO with sign-off.

**Comment by jdemel** [comment]
Thanks for your help! This script is really useful!

**Pull request merged**